Issue/11153 fix file loading error and other issues in PR #11695

ashiagr opened the pull request:
Merge instructions:

1. Review the code and test the changes
2. Make sure #11695 is merged to `feature/media-editing-phase-two `
3. Update target branch to `feature/media-editing-phase-two `
4. Merge the PR

This PR fixes below issues mentioned in #11695 as part of #11153

- [x] Fixes file loading error and displays the error mentioned in #11695 description:

> When the PreviewImage flow is started with images saved in local storage, Glide's asFile() method returns an error - nothing happens after you click on "Crop" action in the Toolbar. We probably shouldn't even call asFile() on local images, wdyt? Moreover, we probably should show a snackbar and show an error to the user when asFile fails.

- [x] Moves numberOfImages logic to VM as requested [here](https://github.com/wordpress-mobile/WordPress-Android/pull/11695#discussion_r410845460). cfe8cdf 
- [x] Supports null `lowResImageUrl` as asked [here](https://github.com/wordpress-mobile/WordPress-Android/pull/11695#discussion_r410844619). 94ca331

    Please consider making necessary changes here: 
    https://github.com/wordpress-mobile/WordPress-Android/commit/39074c2c389829057318343b297a3d7d364d469b#diff-681fa96c5b5159766b71b0b1aa778ca6R534-R536.    Changed here: 9d7452e


It also fixes these additional issues:
- [x] Crop result propagation issue (cropped result for "last selected image" was set on navigating back  from crop screen without cropping the "current selected image") 03bb133
- [x] Load to file issue on trying to crop an already cropped image 8c4ad7a

Note:
Tests are being included in another PR #11690.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Show error when file loading fails

ashiagr committed:
Pass Uri to load file

Allow loading a file either using a file path obtained from the media store (for local images), or using Glide's disk cache (for remote images). Using Uri for this purpose allowing content and remote Uris to be interchangeable.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/editor/ImageEditorInitializer.kt:
```diff
@@ -58,17 +60,22 @@ class ImageEditorInitializer {
         private fun <T : Any> onLoadFailed(model: Any?, listener: ImageEditor.RequestListener<T>, e: Exception?) =
+                listener.onLoadFailed(e, getResourcePath(model))
+
+        private fun getResourcePath(model: Any?): String {
```
Could you please elaborate why we need to get the actual path here? I think it's considered a bad practice - see https://commonsware.com/blog/2016/03/15/how-consume-content-uri.html for more details. I tried to remove the code and the app still seemed to work. However, I haven't done thorough tests.

ashiagr replied:
I was trying to be safe and consistent with the local file uris usage and found actual path being used in other parts of the app like here :

https://github.com/wordpress-mobile/WordPress-Android/blob/36a05e94de9f0e2632000cb26ec7f69d131435b2/WordPress/src/main/java/org/wordpress/android/ui/main/MeFragment.java#L372-L379

I tested and haven't yet found a case when app doesn't work without it. Also we don't have such a conversion when we pass the uris to the preview fragment here: 

https://github.com/wordpress-mobile/WordPress-Android/blob/80c39533a3fb4eb08db616927583dc7cb1cab562/WordPress/src/main/java/org/wordpress/android/ui/photopicker/PhotoPickerFragment.java#L688-L691

So reverting it here: 1a8b0cb

malinajirka replied:
I see, thanks for the clarification!

ashiagr committed:
Revert Uri to actual path conversion

ashiagr committed:
Move numberOfImages to VM

ashiagr committed:
Supports null lowResImageUrl

We don't need to load preview image thumbnail for local images

ashiagr committed:
Fix minor retry layout visibility issue

ashiagr committed:
Remove TODO

ashiagr committed:
Minor fix

ashiagr committed:
Format code

ashiagr committed:
Fix output directory

ashiagr committed:
Fix crop result propagation issue

https://developer.android.com/guide/navigation/navigation-programmatic#returning_a_result

If you’d only like to handle a result only once, you must call remove() on the SavedStateHandle to clear the result. If you do not remove the result, the LiveData will continue to return the last result to any new Observer instances.

ashiagr committed:
Don't call asFile() for a file url received from the crop result

Calling asFile() for a file url results in a resource load error.

Similarly, passing file "urlString" as inputFilePath to ucrop results in an error.
Passing file "path" instead if it is a file url.

ashiagr committed:
Merge branch 'issue/11153-insert-multiple-edited-images-into-gutenberg' into issue/11153-fix-file-loading-error

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/photopicker/PhotoPickerFragment.java

ashiagr committed:
Fix unused import

malinajirka reviewed (commented):
Thanks @ashiagr! I've tested the changes and it works as expected. I've left one comment and one related question. Let me know what you think ;), thanks!

ashiagr committed:
Move file path logic to VM

ashiagr commented on libs/image-editor/ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/preview/PreviewImageFragment.kt:
```diff
@@ -257,9 +281,15 @@ class PreviewImageFragment : Fragment() {
         // TODO: Temporarily added if check to fix this occasional crash
         // https://stackoverflow.com/q/51060762/193545
         if (navController.currentDestination?.id == R.id.preview_dest) {
+            val finalInputFilePath = if (URLUtil.isFileUrl(inputFilePath)) {
```
Hmm. They seem unnecessary after changes to `onCropMenuClicked` so removed here: f3886d8.

ashiagr commented on libs/image-editor/ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/preview/PreviewImageFragment.kt:
```diff
@@ -275,7 +305,17 @@ class PreviewImageFragment : Fragment() {
+            highResImageUrl
+        }
+
+        viewModel.onCropMenuClicked(isFileUrl, finalImageUrl)
```
Moved here f3886d8

malinajirka reviewed (approved):
LGTM, thanks!
